sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -231,6 +231,15 @@ object TypeCoercion {
       })
   }
 
+  /**
+   * Similar to [[findTightestCommonType]] but with string promotion.
+   */
+  def findWiderTypeForTwoExceptDecimals(t1: DataType, t2: DataType): Option[DataType] = {
```

Comment:
Why except Decimals?

Reply:
If we have maps with decimals of different precision as keys. ```Cast``` will fail in analysis phase since it can't cast a key to nullable (potential lost of precision). IMHO, the type mismatch exception from this function will be more accurate. WDYT?

Reply:
Ah, I see, good catch! But it led me to another issue. We can't choose those types possibly to be null as a map key. Instead of adding the method, how about modifying `findTypeForComplex` as something like:

```scala
private def findTypeForComplex(
      t1: DataType,
      t2: DataType,
      findTypeFunc: (DataType, DataType) => Option[DataType]): Option[DataType] = (t1, t2) match {
  ...
    case (MapType(kt1, vt1, valueContainsNull1), MapType(kt2, vt2, valueContainsNull2)) =>
      findTypeFunc(kt1, kt2)
        .filter(kt => !Cast.forceNullable(kt1, kt) && !Cast.forceNullable(kt2, kt))
        .flatMap { kt =>
          findTypeFunc(vt1, vt2).map { vt =>
            MapType(kt, vt, valueContainsNull1 || valueContainsNull2)
          }
      }
  ...
}
```

We might need to have another pr to discuss this.

cc @cloud-fan @gatorsmile

Reply:
On second thoughts, do we really need those? Seems like the current coercions rules don't contain possibly cast to null?

Reply:
ok, I see that this is a matter of `findTypeForComplex`. I'll submit another pr later. Maybe we can go back to `findWiderTypeForTwo` in `TypeCoercion` and `findCommonTypeDifferentOnlyInNullFlag` for `keyType`.

Reply:
I submitted a pr #22086.

Reply:
Thanks for both your PRs! I will submit changes once they get in.